modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusState.scala

```diff
@@ -61,3 +88,10 @@ object ProposalInfo {
   implicit def showInstance[A]: Show[ProposalInfo[A]] = pi =>
     s"ProposalInfo{proposalArtifact=***, proposalArtifactHash=${pi.proposalArtifactHash.show}}"
 }
+
+@derive(eqv, show)
+sealed trait LockStatus
+
+case object Open extends LockStatus
+case object Locked extends LockStatus
+case object Unlocked extends LockStatus
```

Comment:
As we discussed there is a fixed path of state transitions so for example we can't go from `Locked` to `Open`. To make it more bulletproof we could prevent wrong state transitions by using `IndexedStateT ` and secure that on type-level:

```scala
def lock: IndexedStateT[Eval, Open.type, Locked.type, Unit] = IndexedStateT.set(Locked)
def unlock: IndexedStateT[Eval, Locked.type, Unlocked.type, Unit] = IndexedStateT.set(Closed)
```

then if you do 

```
val check = for {
  _ <- lock
  _ <- unlock
  _ <- lock
} yield ()
```

then it gives you a type error.

Moreover if you do `valid.run(Open)` then it is fine because `_ <- lock` can be executed on `Open` but `valid.run(Unlocked)` also gives you a type error.

Reply:
Ok, I tried to include this, but it would require a bigger refactoring, since we would have to include the type of the lock in the type parameters of the whole state `ConsensusState[Key, Artifact, LockStatus]`. This in turn would require changes to all methods that operate on a consensus state (not only method signature would change, but implementation would have to be adjusted in a couple of cases as well).